### DISCUSSION TITLE
WINUI 3 Dependency

### DISCUSSION THREAD

**Discussion by rfuzzo** [microsoft/winget-pkgs#88963] [Q&A] [answered]
How can I submit WINUI 3 apps to winget? The installer works just fine on my or other systems.
I read that dependencies are not supported yet.

https://github.com/microsoft/winget-pkgs/pull/88879#issuecomment-1318725471

Validation errors point to the windowsAppRuntime not being installed.

Thank you :D

```
2022-11-17 06:30:57.429 [CORE] Deployment operation #0: Windows cannot install package 26499Wolvenkit.WolvenKit.Installer_0.1.5.0_x64__str4xaare4n04 because this package depends on a framework that could not be found. 

Provide the framework "Microsoft.WindowsAppRuntime.1.2" published by "CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US", with neutral or x64 processor architecture and minimum version 2000.677.1750.0, along with this package to install. 

The frameworks with name "Microsoft.WindowsAppRuntime.1.2" currently installed are: {}

2022-11-17 06:30:57.429 [FAIL] D:\a\_work\1\s\external\pkg\src\AppInstallerCommonCore\Deployment.cpp(159)\WindowsPackageManager.dll!00007FF94CBB990F: (caller: 00007FF94CB2FA5E) Exception(1) tid(d5c) 80073CF3 Package failed updates, dependency or conflict validation.
```

**Comment by denelon** [comment]
We need to fix:
* https://github.com/microsoft/winget-cli/issues/2202

Then when support for dependencies is stable, this should allow us to take these kinds of packages.

**Reply by rfuzzo** [reply]
Thank you for your reply!. Just to be clear: you're saying that winget currently does not support WINUI 3 packages?

**Reply by denelon** [reply]
That's correct for now unless the dependencies are bundled with the package.

**Comment by rfuzzo** [comment]
Thank you. How would I bundle `Microsoft.WindowsAppRuntime.1.2` with the package?

**Reply by stephengillie** [reply]
The dependency has just become available through Winget. Please add the dependency in your installer.yaml file: 
```
Dependencies:
  PackageDependencies:
    - PackageIdentifier: Microsoft.WindowsAppRuntime.1.2
Installers:
```

**Comment by stephengillie** [comment]
I've found `Microsoft.WindowsAppRuntime.1.2` is available through the MS Store as `9P7VS8XHG9G5`. 

```
Name                  Id           Version Source
---------------------------------------------------
WindowsAppRuntime.1.2 9P7VS8XHG9G5 Unknown msstore
```

Edit: in theory it should work, but in testing it only results in errors, as explained by @dnelon below.

**Reply by denelon** [reply]
This will not work for packages in the community repository. The dependencies must come from the same source.

**Comment by rfuzzo** [comment, marked as answer]
The solution for now is to publish as self-contained

To publish as self-contained use: `dotnet publish <PROJECT> -o publish -c Release --self-contained true -r win10-x64 -p:Platform=x64` (https://github.com/microsoft/WindowsAppSDK/discussions/3026)

and add `<EnablePreviewMsixTooling>true</EnablePreviewMsixTooling>` in the projects (https://github.com/dotnet/maui/issues/5886)
